Abstracting outbound side of transport

VachaShah opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Coming from https://github.com/opensearch-project/OpenSearch/pull/13178#discussion_r1569221067 and as a next step to #12967, abstracting and adding support for multiple protocols on outbound side of transport.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
Part of #10684 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

VachaShah commented:
@reta @dblock Let me know what you guys think.

reta commented on server/src/main/java/org/opensearch/transport/OutboundHandler.java:
```diff
@@ -83,14 +74,16 @@ final class OutboundHandler {
         this.statsTracker = statsTracker;
         this.threadPool = threadPool;
-        this.bigArrays = bigArrays;
+        this.protocolMessageHandlers = Map.of(
```
@VachaShah I think we need to think a bit more how separate that. Here is the idea which I think should work. 

The `OutboundHandler` should be in sync with `InboundHandler` with respect to the protocol: if the inbound message is in protobuf, than the outbound should also be in protocol. There are 2 issues with `OutboundHandler` (at least, as I see it):
 - it is created before `InboundHandler` (so we cannot make those agree on protocol)
 - `OutboundHandler` is mixing 2 things now: sending raw bytes and transport requests / responses (which involves serialization that we would need to change)

The first thing we need to do I believe is to strip from `OutboundHandler` everything which is related to transport requests / responses: basically that would leave as with `void sendBytes(TcpChannel channel, BytesReference bytes, ActionListener<Void> listener)` only. 

Now, the `OutboundHandler` is passed to `InboundHandler` and consequently to `NativeMessageHandler` for now (and protobuf later). At this point things become more straightforward:
 - `NativeMessageHandler` would create an instance of `NativeOutboundHandler`  that would implement the  `sendResponse` / `sendErrorResponse` / `sendRequest` using native protocol and send out raw bytes using the  `OutboundHandler`
 -  consequently, `ProtoobufMessageHandler` would create an instance of `ProtoobufOutboundHandler`  that would implement the  `sendResponse` / `sendErrorResponse` / `sendRequest` using protobuf protocol and send out raw bytes using the  `OutboundHandler`

I think that should work, what do you think?

VachaShah replied:
Interesting! I always thought that `InboundHandler` should be in sync with `OutboundHandler`, so if the outbound message is in protobuf, inbound should also be protobuf.

When we strip everything from `OutboundHandler` except `sendBytes`, are we essentially making `TransportResponse`, `TransportRequest` specific to native protocol? And when protobuf is introduced, it would have its own version of `TransportResponse`?

reta replied:
> Interesting! I always thought that InboundHandler should be in sync with OutboundHandler

Correct, but inbound comes first , right? (this is the request side)

> are we essentially making TransportResponse, TransportRequest specific to native protocol? And when protobuf is introduced, it would have its own version of TransportResponse?

The serdes of TransportRequest  / TransportResponse is protocol specific, yes.

VachaShah replied:
Makes sense! Thank you @reta. Apologies for the delay in response, I was out over the weekend. Let me look into how we can achieve the above idea and I will add an update on this PR once that is done.

VachaShah replied:
Just an update here: `sendBytes` method in `OutboundHandler` is not used by general messages, it is made specially for keep alive messages I think because for `TransportResponse` and such messages, they are serialized to bytes using `MessageSerializer` and sent over the wire so thats a separate flow.

VachaShah replied:
I was looking into this implementation and the problem I faced is that the `OutboundHandler` and `InboundHandler` combination is decided at [`TcpTransport`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TcpTransport.java#L191-L208) class itself. Now I could make `OutboundHandler` as an interface which `NativeOutboundMessageHandler` would implement and the `NativeMessageHandler` can create an instance of this specific native outbound handler ([here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/NativeMessageHandler.java#L95)). But the problem lies in `TcpTransport` where the outbound handler is also used. So we would need a specific implementation here as well and not an interface. Also, as I mentioned in the previous comment, `sendBytes` is misleading, that is used for [keepAlive type messages](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TcpTransport.java#L207), for all other messages (TransportRequest, TransportResponse etc.), the messages are sent using [`internalSend`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/OutboundHandler.java#L195). Before caling `internalSend`, the message will be [serialized](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/OutboundHandler.java#L189) which should have different implementations for each protocol, hence the current code in this PR.

VachaShah replied:
@reta Let me know what you think of the above.

reta replied:
I have created the draft pull request to demonstrate the idea, it makes the `OutboundHandler` context dependant (on the protocol): https://github.com/opensearch-project/OpenSearch/pull/13452

VachaShah replied:
Thank you very much @reta for the draft PR! This makes it very clear! I have incorporated all those changes in this PR :)

reta commented on server/src/main/java/org/opensearch/transport/BytesTransportRequest.java:
```diff
@@ -47,7 +47,7 @@
 public class BytesTransportRequest extends TransportRequest {
 
-    BytesReference bytes;
+    public BytesReference bytes;
```
Please remove `public `

reta commented on server/src/main/java/org/opensearch/transport/CompressibleBytesOutputStream.java:
```diff
@@ -58,13 +58,13 @@
  * @opensearch.internal
  */
-final class CompressibleBytesOutputStream extends StreamOutput {
+public final class CompressibleBytesOutputStream extends StreamOutput {
```
This class could be kept package private (unchanged) and moved to `nativeprotocol` package: it is used only by native protocol

reta commented on server/src/main/java/org/opensearch/transport/NetworkMessage.java:
```diff
@@ -48,7 +48,7 @@ public abstract class NetworkMessage {
     protected final byte status;
 
-    NetworkMessage(ThreadContext threadContext, Version version, byte status, long requestId) {
+    public NetworkMessage(ThreadContext threadContext, Version version, byte status, long requestId) {
```
This constructor could be kept package private (unchanged) and moved to `nativeprotocol` package: it is used only by native protocol

reta commented on server/src/main/java/org/opensearch/transport/ProtocolOutboundHandler.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public abstract class ProtocolOutboundHandler {
+
+    public abstract void sendRequest(
```
Please add javadocs

reta commented on server/src/main/java/org/opensearch/transport/ProtocolOutboundMessage.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @opensearch.internal
+ */
+public interface ProtocolOutboundMessage {
```
This interface is not used and could be removed for now

reta commented on server/src/main/java/org/opensearch/transport/TcpTransport.java:
```diff
@@ -238,7 +254,7 @@ protected void doStart() {}
     @Override
     public synchronized void setMessageListener(TransportMessageListener listener) {
-        outboundHandler.setMessageListener(listener);
+        // outboundHandler.setMessageListener(listener);
```
```suggestion
        handshakerHandler.setMessageListener(listener);
```

reta commented on server/src/main/java/org/opensearch/transport/nativeprotocol/NativeOutboundMessage.java:
```diff
@@ -48,16 +55,17 @@
  * @opensearch.internal
  */
-abstract class OutboundMessage extends NetworkMessage {
+public abstract class NativeOutboundMessage extends NetworkMessage implements ProtocolOutboundMessage {
```
```suggestion
abstract class NativeOutboundMessage extends NetworkMessage implements ProtocolOutboundMessage {
```

VachaShah replied:
It is needed for tests like InboundPipelineTests etc.

VachaShah replied:
Not really since InboundPipeline is common for all protocols, right now it uses native protocol things since it is default and would need references from that.

reta replied:
>  right now it uses native protocol things since it is default and would need references from that.

I think it perfectly justifies the move, for protobuf we will add another test to `protobuf` package very likely

VachaShah replied:
That makes sense! Actually there are quite a few other classes that need to be moved to nativeprotocol package like InboundAggregator, InboundDecoder etc. They kind of go hand in hand with the tests. Do you mind if I open a new PR after this PR for them? It might involve more files.

reta replied:
> Do you mind if I open a new PR after this PR for them? It might involve more files.

Sure, it is up to you, I think the moving is simple and having clean change would definitely be better I think

VachaShah replied:
Sure! Let me fix the 2 tests that are failing and I can add this change as well.

reta commented on server/src/main/java/org/opensearch/transport/nativeprotocol/NativeOutboundMessage.java:
```diff
@@ -48,16 +55,17 @@
 
-    BytesReference serialize(BytesStreamOutput bytesStream) throws IOException {
+    @Override
+    public BytesReference serialize(BytesStreamOutput bytesStream) throws IOException {
```
Please revert to package protected, it is not yet needed

VachaShah replied:
It is needed for tests like InboundPipelineTests etc.

reta commented on server/src/main/java/org/opensearch/transport/OutboundHandler.java:
```diff
@@ -35,164 +35,56 @@
-        ActionListener<Void> listener = ActionListener.wrap(() -> messageListener.onResponseSent(requestId, action, response));
-        sendMessage(channel, message, listener);
-    }
+    // public void setMessageListener(TransportMessageListener listener) {
```
Please remove

reta commented on server/src/main/java/org/opensearch/transport/OutboundHandler.java:
```diff
@@ -35,164 +35,56 @@
-    }
-
-    private void internalSend(TcpChannel channel, SendContext sendContext) throws IOException {
+    public void sendInternalBytes(TcpChannel channel, SendContext sendContext) throws IOException {
```
Let's use a bit better name:

```suggestion
    public void sendBytes(TcpChannel channel, SendContext sendContext) throws IOException {
```

reta commented on server/src/main/java/org/opensearch/transport/ProtocolMessageHandler.java:
```diff
@@ -24,4 +24,6 @@ public void messageReceived(
         TransportMessageListener messageListener
     ) throws IOException;
+
+    public void setMessageListener(TransportMessageListener listener);
```
Please add javadocs

VachaShah commented:
Thank you for the review @reta! I will work on addressing the comments and I am also looking at some test failures which are likely due to message listeners:

[org.opensearch.transport.nio.SimpleMockNioTransportTests.testMessageListeners](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.nio/SimpleMockNioTransportTests/testMessageListeners/)
[org.opensearch.transport.nio.SimpleMockNioTransportTests.testTracerLog](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.nio/SimpleMockNioTransportTests/testTracerLog/)
[org.opensearch.transport.netty4.SimpleNetty4TransportTests.testTracerLog](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.netty4/SimpleNetty4TransportTests/testTracerLog/)
[org.opensearch.transport.netty4.SimpleNetty4TransportTests.testMessageListeners](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.netty4/SimpleNetty4TransportTests/testMessageListeners/)
[org.opensearch.transport.netty4.ssl.SimpleSecureNetty4TransportTests.testTracerLog](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.netty4.ssl/SimpleSecureNetty4TransportTests/testTracerLog/)
[org.opensearch.transport.netty4.ssl.SimpleSecureNetty4TransportTests.testMessageListeners](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.netty4.ssl/SimpleSecureNetty4TransportTests/testMessageListeners/)
[org.opensearch.transport.nio.SimpleNioTransportTests.testTracerLog](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.nio/SimpleNioTransportTests/testTracerLog/)
[org.opensearch.transport.nio.SimpleNioTransportTests.testMessageListeners](https://build.ci.opensearch.org/job/gradle-check/37998/testReport/junit/org.opensearch.transport.nio/SimpleNioTransportTests/testMessageListeners/)

reta commented:
> Thank you for the review @reta! I will work on addressing the comments and I am also looking at some test failures which are likely due to message listeners:

Thanks @VachaShah may be fixed by https://github.com/opensearch-project/OpenSearch/pull/13293#discussion_r1584794718

VachaShah commented:
> > Thank you for the review @reta! I will work on addressing the comments and I am also looking at some test failures which are likely due to message listeners:
> 
> Thanks @VachaShah may be fixed by [#13293 (comment)](https://github.com/opensearch-project/OpenSearch/pull/13293#discussion_r1584794718)

Thank you @reta! Yup that fixed the message listener related tests. I have 2 more test failures to go for InboundHandlerTests which I am looking at:

```
 2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.transport.InboundHandlerTests.testRequestFullyReadButMoreDataIsAvailable" -Dtests.seed=E2FAF4B5C492BE4B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de -Dtests.timezone=Africa/Luanda -Druntime.java=21
  2> java.lang.AssertionError: 
    Expected: an instance of java.lang.IllegalStateException
         but: null
        at __randomizedtesting.SeedInfo.seed([E2FAF4B5C492BE4B:2E02C61843C887C0]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:964)
        at org.junit.Assert.assertThat(Assert.java:930)
        at org.opensearch.transport.InboundHandlerTests.testRequestFullyReadButMoreDataIsAvailable(InboundHandlerTests.java:512)
```

```
2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.transport.InboundHandlerTests.testRequestNotFullyRead" -Dtests.seed=E2FAF4B5C492BE4B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de -Dtests.timezone=Africa/Luanda -Druntime.java=21
  2> java.lang.AssertionError: 
    Expected: an instance of java.lang.IllegalStateException
         but: null
        at __randomizedtesting.SeedInfo.seed([E2FAF4B5C492BE4B:4FCE7D97501DDC2C]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:964)
        at org.junit.Assert.assertThat(Assert.java:930)
        at org.opensearch.transport.InboundHandlerTests.testRequestNotFullyRead(InboundHandlerTests.java:441)

```

VachaShah commented:
> > > Thank you for the review @reta! I will work on addressing the comments and I am also looking at some test failures which are likely due to message listeners:
> > 
> > 
> > Thanks @VachaShah may be fixed by [#13293 (comment)](https://github.com/opensearch-project/OpenSearch/pull/13293#discussion_r1584794718)
> 
> Thank you @reta! Yup that fixed the message listener related tests. I have 2 more test failures to go for InboundHandlerTests which I am looking at:
> 
> ```
>  2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.transport.InboundHandlerTests.testRequestFullyReadButMoreDataIsAvailable" -Dtests.seed=E2FAF4B5C492BE4B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de -Dtests.timezone=Africa/Luanda -Druntime.java=21
>   2> java.lang.AssertionError: 
>     Expected: an instance of java.lang.IllegalStateException
>          but: null
>         at __randomizedtesting.SeedInfo.seed([E2FAF4B5C492BE4B:2E02C61843C887C0]:0)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
>         at org.junit.Assert.assertThat(Assert.java:964)
>         at org.junit.Assert.assertThat(Assert.java:930)
>         at org.opensearch.transport.InboundHandlerTests.testRequestFullyReadButMoreDataIsAvailable(InboundHandlerTests.java:512)
> ```
> 
> ```
> 2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.transport.InboundHandlerTests.testRequestNotFullyRead" -Dtests.seed=E2FAF4B5C492BE4B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de -Dtests.timezone=Africa/Luanda -Druntime.java=21
>   2> java.lang.AssertionError: 
>     Expected: an instance of java.lang.IllegalStateException
>          but: null
>         at __randomizedtesting.SeedInfo.seed([E2FAF4B5C492BE4B:4FCE7D97501DDC2C]:0)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
>         at org.junit.Assert.assertThat(Assert.java:964)
>         at org.junit.Assert.assertThat(Assert.java:930)
>         at org.opensearch.transport.InboundHandlerTests.testRequestNotFullyRead(InboundHandlerTests.java:441)
> ```

Was able to fix this. Will update the PR with required changes.

VachaShah commented:
@reta This is ready for review with the [remaining comment](https://github.com/opensearch-project/OpenSearch/pull/13293#discussion_r1585495910) addressed. I have also split/refactored tests and moved classes and tests if protocol dependent.

reta commented on server/src/main/java/org/opensearch/transport/Header.java:
```diff
@@ -60,11 +60,11 @@ public class Header {
     // These are directly set by tests
-    String actionName;
-    Tuple<Map<String, String>, Map<String, Set<String>>> headers;
+    public String actionName;
```
@VachaShah could you please revert changes in this file and move `InboundDecoder` and `InboundAggregator` back to `transport` package (at least for now)?

Yes, that would impact tests but we need to be a little bit more creative at the moment in order to contain the change.

For tests, the idea is very simple (we would need to move back `InboundDecoderTests`, `InboundPipelineTests` and `InboundAggregatorTests`, and probably `NativeOutboundHandlerTests` as well). If you look at `InboundDecoderTests` and `InboundPipelineTests`, those do not need access to message at all, so we could make these classes abstract and implement for each protocol, for example:

```java
public abstract class InboundDecoderTests extends OpenSearchTestCase {

    protected abstract BytesReference serialize(boolean isRequest, Version version, boolean handshake, boolean compress, 
          String action, long requestId) throws IOException;

     // all tests stay there

     public void testDecode() throws Exception { 
         // ...
         // Instead of creating NativeOutboundMessage directly, we use `serialize` method, that each protocol would implement
         final BytesReference totalBytes = serialize(isRequest, Version.CURRENT, false, false, action, requestId);
          // ...
     }
}
```


```java
package org.opensearch.transport.nativeprotocol;

import org.opensearch.Version;
import org.opensearch.common.io.stream.BytesStreamOutput;
import org.opensearch.core.common.bytes.BytesReference;
import org.opensearch.transport.InboundDecoderTests;
import org.opensearch.transport.TestRequest;
import org.opensearch.transport.TestResponse;

import java.io.IOException;
import java.util.Collections;

public class NativeInboundDecoderTests extends InboundDecoderTests {
    @Override
    protected BytesReference serialize(boolean isRequest, Version version, boolean handshake, boolean compress, String action, long requestId) throws IOException {
        NativeOutboundMessage message;
        if (isRequest) {
            message = new NativeOutboundMessage.Request(
                threadContext,
                new String[0],
                new TestRequest(randomAlphaOfLength(100)),
                version,
                action,
                requestId,
                handshake,
                compress
            );
        } else {
            message = new NativeOutboundMessage.Response(
                threadContext,
                Collections.emptySet(),
                new TestResponse(randomAlphaOfLength(100)),
                version,
                requestId,
                handshake,
                compress
            );
        }

        final BytesReference totalBytes = message.serialize(new BytesStreamOutput());
        return totalBytes;
    }

}

```

reta commented on server/src/test/java/org/opensearch/transport/InboundDecoderTests.java:
```diff
@@ -39,19 +39,29 @@
+public abstract class InboundDecoderTests extends OpenSearchTestCase {
 
-    private ThreadContext threadContext;
+    public ThreadContext threadContext;
```
```suggestion
    protected ThreadContext threadContext;
```

reta commented on server/src/test/java/org/opensearch/transport/InboundDecoderTests.java:
```diff
@@ -65,36 +75,15 @@ public void testDecode() throws IOException {
-            );
+            totalBytes = serialize(isRequest, Version.CURRENT, false, false, action, requestId, new TestResponse(randomAlphaOfLength(100)));
         }
 
```
final BytesReference totalBytes = serialize(isRequest, Version.CURRENT, false, false, action, requestId, new TestResponse(randomAlphaOfLength(100)));

reta commented on server/src/test/java/org/opensearch/transport/InboundDecoderTests.java:
```diff
@@ -65,36 +75,15 @@ public void testDecode() throws IOException {
+        final BytesReference totalBytes;
         if (isRequest) {
             threadContext.putHeader(headerKey, headerValue);
+            totalBytes = serialize(isRequest, Version.CURRENT, false, false, action, requestId, new TestRequest(randomAlphaOfLength(100)));
```
You don't need this conditional logic for `serialize` - these calls are identical,  just use it the same way after `if` block (https://github.com/opensearch-project/OpenSearch/pull/13293/files#r1593249748)

reta commented on server/src/test/java/org/opensearch/transport/InboundDecoderTests.java:
```diff
@@ -183,34 +170,16 @@ public void testCompressedDecode() throws IOException {
-                false,
-                true
-            );
+            totalBytes = serialize(true, Version.CURRENT, false, true, action, requestId, transportMessage);
```
Same here, no need for conditional, after `if`:

```
final BytesReference totalBytes =serialize(isRequest, Version.CURRENT, false, true, action, requestId, transportMessage);
```

reta commented on server/src/test/java/org/opensearch/transport/InboundDecoderTests.java:
```diff
@@ -65,36 +75,16 @@ public void testDecode() throws IOException {
         long requestId = randomNonNegativeLong();
         final String headerKey = randomAlphaOfLength(10);
         final String headerValue = randomAlphaOfLength(20);
+        TransportMessage transportMessage;
```
```suggestion
        final TransportMessage transportMessage = new TestResponse(randomAlphaOfLength(100));;
```

reta commented on server/src/test/java/org/opensearch/transport/InboundHandlerTests.java:
Could we follow the same approach as `InboundDecoderTests` to have `abstract serialize` method?

VachaShah replied:
Sure!

reta commented on server/src/test/java/org/opensearch/transport/TransportLoggerTests.java:
```diff
@@ -93,17 +94,14 @@ public void testLoggingHandler() throws Exception {
         boolean compress = randomBoolean();
         try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
-            OutboundMessage.Request request = new OutboundMessage.Request(
+            return NativeOutboundMessageTests.serializeNativeOutboundRequest(
```
Could we also follow the same approach as `InboundDecoderTests` with `abstract serialize` method?

VachaShah replied:
I think its not needed for `TransportLoggerTests` since serialization is not the part that it is testing. The implementation being tested is for `logInboundMessage` and `logOutboundMessage` which just take in `BytesReference`. The native request is just being used to create bytes. WDYT?

reta replied:
The idea behind `serialize` (or alternative) is to abstract over the protocol, how the `TransportLoggerTests` would deal with protobuf message? (since it directly uses `NativeOutboundMessageTests` here). To your point, `serialize` is really not a good fir here, but making this method `private BytesReference buildRequest() throws IOException` would be much better option.

VachaShah replied:
Makes sense!

dblock commented:
Looks great, let's merge it.

dblock commented:
@VachaShah manual backport pls :(
